File: backend/routes/hospitals.js
import { Router } from 'express';
import Hospital from '../models/Hospital.js';
import { verifyToken } from '../middleware/authMiddleware.js';

const router = Router();

// Get current hospital profile
router.get('/me', verifyToken, async (req, res) => {
    try {
        const hospital = await Hospital.getByUserId(req.user.id);
        if (!hospital) {
            return res.status(404).json({ success: false, error: 'Hospital profile not found' });
        }
        res.json({ success: true, data: hospital });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Get all hospitals
router.get('/', async (req, res) => {
    try {
        const filters = {
            city: req.query.city,
            search: req.query.search
        };

        const hospitals = await Hospital.getAll(filters);
        res.json({ success: true, data: hospitals });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Get hospital statistics
router.get('/stats', async (req, res) => {
    try {
        const stats = await Hospital.getStats();
        res.json({ success: true, data: stats });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Get single hospital
router.get('/:id', async (req, res) => {
    try {
        const hospital = await Hospital.getById(req.params.id);
        if (!hospital) {
            return res.status(404).json({ success: false, error: 'Hospital not found' });
        }
        res.json({ success: true, data: hospital });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Create hospital
router.post('/', async (req, res) => {
    try {
        const { name, address, city, phone, email, latitude, longitude, emergency_contact } = req.body;

        if (!name || !address || !city || !phone) {
            return res.status(400).json({
                success: false,
                error: 'Name, address, city, and phone are required'
            });
        }

        const hospital = await Hospital.create({ name, address, city, phone, email, latitude, longitude, emergency_contact });
        res.status(201).json({ success: true, data: hospital });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Update hospital
router.put('/:id', async (req, res) => {
    try {
        const hospital = await Hospital.getById(req.params.id);
        if (!hospital) {
            return res.status(404).json({ success: false, error: 'Hospital not found' });
        }

        const updated = await Hospital.update(req.params.id, req.body);
        res.json({ success: true, data: updated });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Delete hospital
router.delete('/:id', async (req, res) => {
    try {
        const hospital = await Hospital.getById(req.params.id);
        if (!hospital) {
            return res.status(404).json({ success: false, error: 'Hospital not found' });
        }

        await Hospital.delete(req.params.id);
        res.json({ success: true, message: 'Hospital deleted successfully' });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

export default router;

File: backend_python/routes/hospitals.py
"""
Hospital Routes for BEOS Python Backend
"""

from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel
from typing import Optional
from models.hospital import Hospital
from middleware.auth import verify_token

router = APIRouter()


class HospitalCreate(BaseModel):
    name: str
    address: str
    city: str
    phone: str
    email: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    emergency_contact: Optional[str] = None


class HospitalUpdate(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    emergency_contact: Optional[str] = None


@router.get("/me")
async def get_my_profile(current_user: dict = Depends(verify_token)):
    """Get current hospital profile"""
    try:
        hospital = await Hospital.get_by_user_id(current_user["id"])
        if not hospital:
            raise HTTPException(
                status_code=404,
                detail={"success": False, "error": "Hospital profile not found"}
            )
        return {"success": True, "data": hospital}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.get("/")
async def get_hospitals(
    city: Optional[str] = Query(None),
    search: Optional[str] = Query(None)
):
    """Get all hospitals"""
    try:
        filters = {}
        if city:
            filters["city"] = city
        if search:
            filters["search"] = search
        
        hospitals = await Hospital.get_all(filters)
        return {"success": True, "data": hospitals}
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.get("/stats")
async def get_hospital_stats():
    """Get hospital statistics"""
    try:
        stats = await Hospital.get_stats()
        return {"success": True, "data": stats}
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.get("/{hospital_id}")
async def get_hospital(hospital_id: int):
    """Get single hospital"""
    try:
        hospital = await Hospital.get_by_id(hospital_id)
        if not hospital:
            raise HTTPException(
                status_code=404,
                detail={"success": False, "error": "Hospital not found"}
            )
        return {"success": True, "data": hospital}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.post("/")
async def create_hospital(hospital: HospitalCreate):
    """Create hospital"""
    try:
        if not hospital.name or not hospital.address or not hospital.city or not hospital.phone:
            raise HTTPException(
                status_code=400,
                detail={"success": False, "error": "Name, address, city, and phone are required"}
            )
        
        new_hospital = await Hospital.create(hospital.model_dump())
        return {"success": True, "data": new_hospital}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.put("/{hospital_id}")
async def update_hospital(hospital_id: int, hospital: HospitalUpdate):
    """Update hospital"""
    try:
        existing = await Hospital.get_by_id(hospital_id)
        if not existing:
            raise HTTPException(
                status_code=404,
                detail={"success": False, "error": "Hospital not found"}
            )
        
        updated = await Hospital.update(hospital_id, hospital.model_dump(exclude_unset=True))
        return {"success": True, "data": updated}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})


@router.delete("/{hospital_id}")
async def delete_hospital(hospital_id: int):
    """Delete hospital"""
    try:
        existing = await Hospital.get_by_id(hospital_id)
        if not existing:
            raise HTTPException(
                status_code=404,
                detail={"success": False, "error": "Hospital not found"}
            )
        
        await Hospital.delete(hospital_id)
        return {"success": True, "message": "Hospital deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail={"success": False, "error": str(e)})
